path/spanning_tree.go

```diff
@@ -5,101 +5,166 @@
 package path
 
 import (
+	"container/heap"
+	"math"
 	"sort"
 
 	"github.com/gonum/graph"
 	"github.com/gonum/graph/concrete"
-	"github.com/gonum/graph/internal"
 )
 
-// EdgeListerGraph is an undirected graph than returns its complete set of edges.
-type EdgeListerGraph interface {
+// UndirectedWeighter is an undirected graph that returns distinct edge weights.
+type UndirectedWeighter interface {
 	graph.Undirected
-	Edges() []graph.Edge
+	graph.Weighter
 }
 
 // Prim generates a minimum spanning tree of g by greedy tree extension, placing
-// the result in the destination. The destination is not cleared first.
-func Prim(dst graph.MutableUndirected, g EdgeListerGraph) {
-	var weight Weighting
-	if wg, ok := g.(graph.Weighter); ok {
-		weight = wg.Weight
-	} else {
-		weight = UniformCost(g)
+// the result in the destination. The destination is not cleared first. The weight
+// of the minimum spanning tree is returned. If g is not connected, the minimum
+// spanning forest will be constructed in dst and the sum of minimum spanning tree
+// weights will be returned.
+func Prim(dst graph.MutableUndirected, g UndirectedWeighter) float64 {
+	nodes := g.Nodes()
+	if len(nodes) == 0 {
+		return 0
 	}
-
-	nlist := g.Nodes()
-
-	if nlist == nil || len(nlist) == 0 {
-		return
+	u := nodes[0]
+	q := &primQueue{
+		indexOf: make(map[int]int, len(nodes)-1),
+		nodes:   make([]concrete.Edge, 0, len(nodes)-1),
 	}
-
-	dst.AddNode(nlist[0])
-	remain := make(internal.IntSet)
-	for _, node := range nlist[1:] {
-		remain.Add(node.ID())
+	for _, u := range nodes[1:] {
+		heap.Push(q, concrete.Edge{F: u, W: math.Inf(1)})
 	}
+	for _, v := range g.From(u) {
+		w, ok := g.Weight(u, v)
+		if !ok {
+			panic("prim: unexpected invalid weight")
+		}
+		q.update(v, u, w)
+	}
+
+	var w float64
+	for q.Len() > 0 {
+		e := heap.Pop(q).(concrete.Edge)
+		if e.To() != nil && g.HasEdge(e.From(), e.To()) {
+			dst.SetEdge(e)
+			w += e.Weight()
+		}
 
-	edgeList := g.Edges()
-	for remain.Count() != 0 {
-		var edges []concrete.Edge
-		for _, e := range edgeList {
-			u := e.From()
-			v := e.To()
-			if (dst.Has(u) && remain.Has(v.ID())) || (dst.Has(v) && remain.Has(u.ID())) {
-				w, ok := weight(u, v)
+		u = e.From()
+		for _, n := range g.From(u) {
+			if key, ok := q.key(n); ok {
+				w, ok := g.Weight(u, n)
 				if !ok {
 					panic("prim: unexpected invalid weight")
 				}
-				edges = append(edges, concrete.Edge{F: u, T: v, W: w})
+				if w < key {
+					q.update(n, u, w)
+				}
 			}
 		}
-
-		sort.Sort(byWeight(edges))
-		min := edges[0]
-		dst.SetEdge(min)
-		remain.Remove(min.From().ID())
 	}
+	return w
+}
+
+// primQueue is an Prim's priority queue.
+type primQueue struct {
+	indexOf map[int]int
+	nodes   []concrete.Edge
```

Comment:
`Prim()`'s code is almost self-documenting, but it took me time to understand the (key) role of the queue. Even though not exported, I think it would be better to document it a bit more and avoid decyphering the code. What took me most time to sort out is what is this slice (it is called `nodes` but it is a slice of edges), what edges are stored here and what are their From and To nodes (`update()` is called with To, From, but interprets them as From, To).
*